### ISSUE TITLE
New Service Energyforecast.de with 48h (and 96h) forecast

### ISSUE THREAD

**Issue by Adminius** [mampfes/ha_epex_spot#146]
Hey,
there is new service available that predicts prices for nex 48/96 hours.

[energyforecast](https://www.energyforecast.de/)

It would be cool if ha_exex_sport could also support it.

**Comment by Adminius** [comment]
URL: 
https://www.energyforecast.de/api/v1/predictions/next_48_hours?token=___API_KEY____&fixed_cost_cent=14.907&vat=19

Response:
`[{"start":"2024-08-09T00:00:00.000Z","end":"2024-08-09T01:00:00.000Z","price":0.2467},{"start":"2024-08-09T01:00:00.000Z","end":"2024-08-09T02:00:00.000Z","price":0.2431},{"start":"2024-08-09T02:00:00.000Z","end":"2024-08-09T03:00:00.000Z","price":0.23},{"start":"2024-08-09T03:00:00.000Z","end":"2024-08-09T04:00:00.000Z","price":0.2432},{"start":"2024-08-09T04:00:00.000Z","end":"2024-08-09T05:00:00.000Z","price":0.2618},{"start":"2024-08-09T05:00:00.000Z","end":"2024-08-09T06:00:00.000Z","price":0.2718},{"start":"2024-08-09T06:00:00.000Z","end":"2024-08-09T07:00:00.000Z","price":0.2373},{"start":"2024-08-09T07:00:00.000Z","end":"2024-08-09T08:00:00.000Z","price":0.2038},{"start":"2024-08-09T08:00:00.000Z","end":"2024-08-09T09:00:00.000Z","price":0.1773},{"start":"2024-08-09T09:00:00.000Z","end":"2024-08-09T10:00:00.000Z","price":0.1673},{"start":"2024-08-09T10:00:00.000Z","end":"2024-08-09T11:00:00.000Z","price":0.1666},{"start":"2024-08-09T11:00:00.000Z","end":"2024-08-09T12:00:00.000Z","price":0.1664},{"start":"2024-08-09T12:00:00.000Z","end":"2024-08-09T13:00:00.000Z","price":0.1656},{"start":"2024-08-09T13:00:00.000Z","end":"2024-08-09T14:00:00.000Z","price":0.1665},{"start":"2024-08-09T14:00:00.000Z","end":"2024-08-09T15:00:00.000Z","price":0.1666},{"start":"2024-08-09T15:00:00.000Z","end":"2024-08-09T16:00:00.000Z","price":0.1697},{"start":"2024-08-09T16:00:00.000Z","end":"2024-08-09T17:00:00.000Z","price":0.2163},{"start":"2024-08-09T17:00:00.000Z","end":"2024-08-09T18:00:00.000Z","price":0.263},{"start":"2024-08-09T18:00:00.000Z","end":"2024-08-09T19:00:00.000Z","price":0.3154},{"start":"2024-08-09T19:00:00.000Z","end":"2024-08-09T20:00:00.000Z","price":0.307},{"start":"2024-08-09T20:00:00.000Z","end":"2024-08-09T21:00:00.000Z","price":0.3017},{"start":"2024-08-09T21:00:00.000Z","end":"2024-08-09T22:00:00.000Z","price":0.2775},{"start":"2024-08-09T22:00:00.000Z","end":"2024-08-09T23:00:00.000Z","price":0.2679},{"start":"2024-08-09T23:00:00.000Z","end":"2024-08-10T00:00:00.000Z","price":0.2762},{"start":"2024-08-10T00:00:00.000Z","end":"2024-08-10T01:00:00.000Z","price":0.2705},{"start":"2024-08-10T01:00:00.000Z","end":"2024-08-10T02:00:00.000Z","price":0.2674},{"start":"2024-08-10T02:00:00.000Z","end":"2024-08-10T03:00:00.000Z","price":0.2653},{"start":"2024-08-10T03:00:00.000Z","end":"2024-08-10T04:00:00.000Z","price":0.2683},{"start":"2024-08-10T04:00:00.000Z","end":"2024-08-10T05:00:00.000Z","price":0.2722},{"start":"2024-08-10T05:00:00.000Z","end":"2024-08-10T06:00:00.000Z","price":0.257},{"start":"2024-08-10T06:00:00.000Z","end":"2024-08-10T07:00:00.000Z","price":0.2046},{"start":"2024-08-10T07:00:00.000Z","end":"2024-08-10T08:00:00.000Z","price":0.1668},{"start":"2024-08-10T08:00:00.000Z","end":"2024-08-10T09:00:00.000Z","price":0.163},{"start":"2024-08-10T09:00:00.000Z","end":"2024-08-10T10:00:00.000Z","price":0.1521},{"start":"2024-08-10T10:00:00.000Z","end":"2024-08-10T11:00:00.000Z","price":0.1398},{"start":"2024-08-10T11:00:00.000Z","end":"2024-08-10T12:00:00.000Z","price":0.118},{"start":"2024-08-10T12:00:00.000Z","end":"2024-08-10T13:00:00.000Z","price":0.1189},{"start":"2024-08-10T13:00:00.000Z","end":"2024-08-10T14:00:00.000Z","price":0.1313},{"start":"2024-08-10T14:00:00.000Z","end":"2024-08-10T15:00:00.000Z","price":0.1568},{"start":"2024-08-10T15:00:00.000Z","end":"2024-08-10T16:00:00.000Z","price":0.1667},{"start":"2024-08-10T16:00:00.000Z","end":"2024-08-10T17:00:00.000Z","price":0.2247},{"start":"2024-08-10T17:00:00.000Z","end":"2024-08-10T18:00:00.000Z","price":0.2935},{"start":"2024-08-10T18:00:00.000Z","end":"2024-08-10T19:00:00.000Z","price":0.3158},{"start":"2024-08-10T19:00:00.000Z","end":"2024-08-10T20:00:00.000Z","price":0.3249},{"start":"2024-08-10T20:00:00.000Z","end":"2024-08-10T21:00:00.000Z","price":0.3068},{"start":"2024-08-10T21:00:00.000Z","end":"2024-08-10T22:00:00.000Z","price":0.2935},{"start":"2024-08-10T22:00:00.000Z","end":"2024-08-10T23:00:00.000Z","price":0.2325},{"start":"2024-08-10T23:00:00.000Z","end":"2024-08-11T00:00:00.000Z","price":0.2319},{"start":"2024-08-11T00:00:00.000Z","end":"2024-08-11T01:00:00.000Z","price":0.2256},{"start":"2024-08-11T01:00:00.000Z","end":"2024-08-11T02:00:00.000Z","price":0.2184},{"start":"2024-08-11T02:00:00.000Z","end":"2024-08-11T03:00:00.000Z","price":0.2149},{"start":"2024-08-11T03:00:00.000Z","end":"2024-08-11T04:00:00.000Z","price":0.2152},{"start":"2024-08-11T04:00:00.000Z","end":"2024-08-11T05:00:00.000Z","price":0.2435},{"start":"2024-08-11T05:00:00.000Z","end":"2024-08-11T06:00:00.000Z","price":0.2696},{"start":"2024-08-11T06:00:00.000Z","end":"2024-08-11T07:00:00.000Z","price":0.2624},{"start":"2024-08-11T07:00:00.000Z","end":"2024-08-11T08:00:00.000Z","price":0.2488},{"start":"2024-08-11T08:00:00.000Z","end":"2024-08-11T09:00:00.000Z","price":0.2373},{"start":"2024-08-11T09:00:00.000Z","end":"2024-08-11T10:00:00.000Z","price":0.2357},{"start":"2024-08-11T10:00:00.000Z","end":"2024-08-11T11:00:00.000Z","price":0.2271},{"start":"2024-08-11T11:00:00.000Z","end":"2024-08-11T12:00:00.000Z","price":0.229},{"start":"2024-08-11T12:00:00.000Z","end":"2024-08-11T13:00:00.000Z","price":0.2278},{"start":"2024-08-11T13:00:00.000Z","end":"2024-08-11T14:00:00.000Z","price":0.2485}]`

**Comment by mampfes** [comment]
Hi, thanks for the info. The idea is great, but I'm a little bit skeptical because obviously the service is provided by a private person and it is unclear to me how the services is funded.

**Comment by moritzmair** [comment]
Hi @mampfes , I'm the developer of [energyforecast.de](https://www.energyforecast.de/) we can work together to make it work. For now this is funded by me. I will keep the 48 hour forecast for free but will add a subscription for a 96h forecast.

**Comment by heinemannj** [comment]
There is a [fork](https://github.com/moritzmair/ha_epex_spot) which I'm testing since today morning - looks stable and contains really useful staff - big improvement by providing EPEXSpot forcasting for DE-LU.

I'm working very closely with the developer since 1.5 week - trustful support and relationship!!!
I will try to convince him to extend this forecast service to other EPEXSpot areas - step by step ;-)

The other implemented staff is following stricly the new [3.0.0-dev-4](https://github.com/mampfes/ha_epex_spot/releases/tag/3.0.0-dev-4).

My opinion:
- Two points for improvement which I have already addressed
- After that - Ready to MERGE!!!

**Comment by Adminius** [comment]
I'm also using same fork since few days without complains ;)

**Comment by suaveolent** [comment]
@moritzmair @heinemannj  great news. 

I am currently putting the final touches towards releasing 3.0.0 version and addressing some of the open issues. 

Let me know if you need some help.

**Comment by suaveolent** [comment]
Yes 3.0.0 introduced a lot of changes, unfortunately, hence the major version increase.

**Comment by heinemannj** [comment]
**Improvement for the documentation:**

EPEX Spot electricity price forecasts by the usage of [EPEXSpot integration](https://github.com/mampfes/ha_epex_spot).

This custom component adds electricity prices from the European Power EXchange ( [EPEX Spot](https://www.epexspot.com/en/about) ) into HA.
It covers the following countries at this time:

|Country|Market area|[Awattar (de)](https://energy.tado.com/services/api)|[Energyforecast](https://www.energyforecast.de/api-docs/index.html)|[EPEX Spot Web Scraper](https://www.epexspot.com/en/market-results)|[SMARD.de](https://www.smard.de/home)|[smartENERGY.at](https://www.smartenergy.at/api-schnittstellen)|[Tibber](https://developer.tibber.com/explorer)|
|---|---|:---:|:---:|:---:|:---:|:---:|:---:|
|Austria|AT|X||X|X|X|
|Belgium|BE|||X|X|
|Czech Republic|CZ||||X|
|Denmark|DK1|||X|X|
||DK2|||X|X|
|Finnland|FI|||X|
|France|FR|||X|X|
|Germany|DE-LU|X|X|X|X||X|
|Luxemburg|DE-LU|X|X|X|X|
|Hungary|HU||||X|
|Italy (North)|IT (North)||||X|
|Netherland|NL|||X|X||X|
|Norway|NO1|||X|||X|
||NO2|||X|X||X|
||NO3|||X|||X|
||NO4|||X|||X|
||NO5|||X|||X|
|Poland|PL|||X|X|
|Slovenia|SI||||X|
|Sweden|SE1|||X|||X|
||SE2|||X|||X|
||SE3|||X|||X|
||SE4|||X|||X|
|Switzerland|CH|||X|X|
|United Kingdom (Great Britain)|GB|||X|
||GB-30|||X|

**Comment by suaveolent** [comment]
Will add that.

**Comment by moritzmair** [comment]
I worked on the fork. Do you guys need a pull request from my side for merging? I'm happy to help, just let me know.

**Comment by suaveolent** [comment]
Yes a PR would be the preferred way.

**Comment by heinemannj** [comment]
After the fully integration of the forecast data into my dashboards I end with the following additional request for improvement:

- Please add the `price_origin` attribute from the API response into the EPEXSpot sensor's data (e.g. `sensor.epex_spot_data_net_price` and `sensor.epex_spot_data_price`).

Energyforecast API response data:

```
  {
    "start": "2024-12-08T23:00:00.000+01:00",
    "end": "2024-12-09T00:00:00.000+01:00",
    "price": 0.075,
    "price_origin": "market"
  },
  {
    "start": "2024-12-09T00:00:00.000+01:00",
    "end": "2024-12-09T01:00:00.000+01:00",
    "price": 0.13239,
    "price_origin": "forecast"
  },
```

@suaveolent: Can you please support?

**Comment by moritzmair** [comment]
I created the pull request: https://github.com/mampfes/ha_epex_spot/pull/180

**Comment by suaveolent** [comment]
> After the fully integration of the forecast data into my dashboards I end with the following additional request for improvement:
> 
> * Please add the `price_origin` attribute from the API response into the EPEXSpot sensor's data (e.g. `sensor.epex_spot_data_net_price` and `sensor.epex_spot_data_price`).
> 
> Energyforecast API response data:
> 
> ```
>   {
>     "start": "2024-12-08T23:00:00.000+01:00",
>     "end": "2024-12-09T00:00:00.000+01:00",
>     "price": 0.075,
>     "price_origin": "market"
>   },
>   {
>     "start": "2024-12-09T00:00:00.000+01:00",
>     "end": "2024-12-09T01:00:00.000+01:00",
>     "price": 0.13239,
>     "price_origin": "forecast"
>   },
> ```
> 
> @suaveolent: Can you please support?

This would most likely end up somwhere here:

https://github.com/mampfes/ha_epex_spot/blob/410154a41c3fb811e891689d473598bea0744fe6/custom_components/epex_spot/sensor.py#L79

However, it might be a good idea to introduce a generic sensor attribute interface, e.g. `data_specific_attributes`, which can be retrieved for each epex spot api separately. 

Once I have merge the PR from @moritzmair, I'll have a look at that.

**Comment by suaveolent** [comment]
Thanks @moritzmair.

It has been successfully merged and is available with `3.0.0-dev-5`.
